src/fastpath/fastpath.c

```diff
@@ -15,6 +15,268 @@
 #endif
 #include <benchmark/benchmark_utilisation.h>
 
+#ifdef CONFIG_ARCH_ARM
+static inline
+#ifndef CONFIG_ARCH_ARM_V6
+FORCE_INLINE
+#endif
+#endif
+void NORETURN fastpath_signal(word_t cptr, word_t msgInfo)
+{
+    cap_t newVTable;
+    vspace_root_t *cap_pd;
+    pde_t stored_hw_asid;
+    word_t fault_type;
+    dom_t dom;
+
+    /* Get fault type. */
+    fault_type = seL4_Fault_get_seL4_FaultType(NODE_STATE(ksCurThread)->tcbFault);
+
+    /* Check there's no saved fault */
+    if (unlikely(fault_type != seL4_Fault_NullFault)) {
+        slowpath(SysSend);
+    }
+
+    /* We land up in here on every SYSCALL_SEND, so we need to check that we are
+     * actually signalling a notification */
+
+    /* Lookup the cap */
+    cap_t cap = lookup_fp(TCB_PTR_CTE_PTR(NODE_STATE(ksCurThread), tcbCTable)->cap, cptr);
+
+    /* Check it's a notification */
+    if (!cap_capType_equals(cap, cap_notification_cap)) {
+        slowpath(SysSend);
+    }
+
+    /* Check that we are allowed to signal this notification */
+    if (!unlikely(cap_notification_cap_get_capNtfnCanSend(cap))) {
+        slowpath(SysSend);
+    }
+
+    /* Get the notification address */
+    notification_t *ntfnPtr = NTFN_PTR(cap_notification_cap_get_capNtfnPtr(cap));
+
+    /* Get the notification state */
+    uint32_t ntfnState = notification_ptr_get_state(ntfnPtr);
+
+    /* Get the notification badge */
+    word_t badge = cap_notification_cap_get_capNtfnBadge(cap);
+
+    switch (ntfnState) {
+        case NtfnState_Active: {
+            word_t badge2 = notification_ptr_get_ntfnMsgIdentifier(ntfnPtr);
+            badge2 |= badge;
+            notification_ptr_set_ntfnMsgIdentifier(ntfnPtr, badge2);
+            restore_user_context();
+        }
+        case NtfnState_Idle: {
+            /* TODO: If bound thread is higher prio, we should switch to it
+             * directly via switchToThread_fp; fastpath_restore; We shouldn't
+             * need to invoke the scheduler. */
+            tcb_t *tcb = (tcb_t *)notification_ptr_get_ntfnBoundTCB(ntfnPtr);
+            /* Check if we are bound and that thread is waiting for a message */
+            if (tcb) {
+                if (thread_state_ptr_get_tsType(&tcb->tcbState) == ThreadState_BlockedOnReceive) {
+                    /* Send and start thread running */
+                    cancelIPC(tcb);
+                    setThreadState(tcb, ThreadState_Running);
```

Comment:
Can fastpath inline versions of these be defined in fastpath.h ?

Reply:
I'm now using the [`thread_state_ptr_set_tsType_np`](https://github.com/seL4/seL4/blob/master/include/fastpath/fastpath.h#L55-L58) defined in fastpath.h to set the thread state. Do you suggest that I also define an inline version of `cancelIPC` in fastpath.h?

Reply:
Yeah I think so.

Reply:
Cool, I've had a go at implementing a [`cancelIPC_fp`](https://github.com/shanekadish/seL4/blob/signal-fastpath/include/fastpath/fastpath.h#L146-L211) and [`cancelSignal_fp`](https://github.com/shanekadish/seL4/blob/signal-fastpath/include/fastpath/fastpath.h#L122-L144) in fastpath.h. Let me know what you think.